### PR TITLE
Handle @llvm.memset.* with non-constant arguments

### PR DESCRIPTION
**Pull request by AlexeySotkin**
There is no SPIR-V counterpart for @llvm.memset.* intrinsic. Cases with constant
value and length arguments are emulated via "storing" a constant array to the
destination.

For other cases we wrap the intrinsic in @spirv.llvm_memset_* function and
expand the intrinsic to a loop via expandMemSetAsLoop() from
llvm/Transforms/Utils/LowerMemIntrinsics.h.

During reverse translation from SPIR-V to LLVM IR we can detect
@spirv.llvm_memset_* and replace it with @llvm.memset.

Signed-off-by: Alexey Sotkin <alexey.sotkin@intel.com>

### CODE REVIEW

File: lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -2720,21 +2720,33 @@ Function *SPIRVToLLVM::transFunction(SPIRVFunction *BF) {
   auto IsKernel = isKernel(BF);
   auto Linkage = IsKernel ? GlobalValue::ExternalLinkage : transLinkageType(BF);
   FunctionType *FT = dyn_cast<FunctionType>(transType(BF->getFunctionType()));
-  Function *F = cast<Function>(
-      mapValue(BF, Function::Create(FT, Linkage, BF->getName(), M)));
+  std::string FuncName = BF->getName();
+  StringRef FuncNameRef(FuncName);
+  // Transform "@spirv.llvm.memset.p0i8.i32.volatile" to @llvm.memset.p0i8.i32
```
Comment (svenvh) [inline lib/SPIRV/SPIRVReader.cpp:2725]:
```suggestion
  // Transform "@spirv.llvm_memset_p0i8_i32.volatile" to @llvm.memset.p0i8.i32
```

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp [hunk 2]
```diff
@@ -84,6 +94,49 @@ class SPIRVRegularizeLLVM : public ModulePass {
...
+    FuncName += ".volatile";
+
+  // Redirect @llvm.memset.* call to @spirv.llvm_memset_*
+  Function *F = M->getFunction(FuncName);
+  if (F) {
+    // This function is already linked in.
+    MSI->setCalledFunction(F);
+    return;
+  }
+  // TODO copy arguments attributes: nocapture writeonly.
```
Comment (svenvh) [inline lib/SPIRV/SPIRVRegularizeLLVM.cpp:115]: Are you still planning to do this as part of this PR?

File: test/transcoding/llvm.memset.ll [hunk 3]
```diff
@@ -36,14 +63,34 @@ target triple = "spir"
...
+
+  ; non-const value
+  tail call void @llvm.memset.p0i8.i32(i8* align 4 %x.bc, i8 %v, i32 3, i1 false)
+; CHECK-LLVM: call void @llvm.memset.p0i8.i32(i8* %x.bc, i8 %v, i32 3, i1 false)
+
+  ; non-const value and size
+  tail call void @llvm.memset.p0i8.i32(i8*  align 4 %x.bc, i8 %v, i32 %s1, i1 false)
+; CHECK-LLVM: call void @llvm.memset.p0i8.i32(i8* %x.bc, i8 %v, i32 %s1, i1 false)
+
+  ; Address spaces
```
Comment (svenvh) [inline test/transcoding/llvm.memset.ll:83]:
```suggestion
  ; Address spaces, non-const value and size
```

**Review by svenvh — COMMENTED** [review]
LGTM besides some minor comments.

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp (hunk 2 above)
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVRegularizeLLVM.cpp:115]: No, I'd prefer to do it in a separate PR.

**Review by svenvh — APPROVED** [review]

**Pull request merged**